### DISCUSSION TITLE
Is project file .friction binary?

### DISCUSSION THREAD

**Discussion by pgilfernandez** [friction2d/friction#290] [General]
Hi,

Is the `.friction` project file saved in binary?
FOSS usually saves projects as XML or ZIP file with content that is browsable, editable, that is, encourage interoperability.
If no other software but Friction can open a project, aren't we closing the door for exporting/importing the project files into other FOSS?

Cheers

**Comment by rodlie** [comment]
Yes they are binary. v2 will probably move to a zip with a bunch of XML's files.

> If no other software but Friction can open a project, aren't we closing the door for exporting/importing the project files into other FOSS?

Tell that to GIMP ;)

But, yeah... As said this will change at some point, but for other reasons than import/export to other applications. My priority is Friction.

There are also better ways to communicate with other applications if needed (scripting, IPC etc).

**Reply by pgilfernandez** [reply]
I also like the option to create project files by creating them with scripts from scratch (without the need of using the editor), for changing parameters with REGEX, ...

**Reply by rodlie** [reply]
Sure, in a perfect world Friction would have a different file format, but as all projects we have technical debt, the file format being one of them.

v1.x will keep the current file format, this will not change until v2, as it will introduce major breakage. On a positive note the new file format exists and is probably ~60% done.

---

But, what you describe can be done in other ways, with some scripting support. This is something I can implement before v2.

And regarding interoperability, this could also be discussed. Give me a use-case and we can probably figure out something.

Example: I use Natron for tracking, I wrote a simple python addon for Natron that converts the transforms in Natron to JSON and then put it in the clipboard, I can then tab to Friction and paste the data into the expression editor, and I have tracking data in Friction.

**Comment by rodlie** [comment]
Forgot to mention, we have progress on the new file format, currently named `.fdesign`. It's getting close to feature parity with the `.friction` file format.

See `designer` branch for progress.

I will experiment with the new file format in a spinoff project this year, and when complete be merged into `main` as an alternative project format. It will then replace the old file format completely for version 2.0 (and the old file format will be removed, we will deprecate the old file format in the last feature version for v1.x).

**Comment by pgilfernandez** [comment]
Wow!

I love it... I really love the power of text based project files, I'll test it when v1.0 RC1 is out 💪🏻

**Comment by rodlie** [comment]
The file structure will look like this:

test.fdesign (zip)
```
├── document.xml
├── mimetype
├── scenes
│   └── 0
│       ├── gradients.xml
│       ├── objects
│       │   ├── 0
│       │   │   └── properties.xml
│       │   ├── 1
│       │   │   ├── assets
│       │   │   │   └── Text
│       │   │   │       └── value.txt
│       │   │   └── properties.xml
│       │   ├── 2
│       │   │   └── properties.xml
│       │   ├── 3
│       │   │   └── properties.xml
│       │   ├── 4
│       │   │   └── properties.xml
│       │   └── 5
│       │       └── properties.xml
│       ├── properties.xml
│       └── stack.xml
└── UI
    └── layouts.xml
```

document.xml
```xml
<Document format-version="1">
 <ColorBookmarks/>
 <Scenes>
  <Scene frame="0" height="1080" clip="false" frameRange="0 300" width="1920" name="Scene 0" fps="30" resolution="0.5"/>
 </Scenes>
</Document>
```

scenes/0/stack.xml
```xml
<Stack>
 <Object type="0" name="Path 0"/>
 <Object type="4" name="Text"/>
 <Object type="17" name="null object"/>
 <Object type="3" name="RectangleBox"/>
 <Object type="1" name="Circle"/>
 <Object type="0" name="Path"/>
</Stack>
```

scenes/0/objects/0/properties.xml
```xml
<Object id="1">
 <Transform>
  <BeginEvent value="0"/>
  <EndEvent value="0"/>
  <Translation>
   <X value="1583.743387160117"/>
   <Y value="92.406496176681202"/>
  </Translation>
  <Rotation value="0"/>
  <Scale>
   <X value="1"/>
   <Y value="1"/>
  </Scale>
  <Shear>
   <X value="0"/>
   <Y value="0"/>
  </Shear>
  <Pivot>
   <X value="0"/>
   <Y value="0"/>
  </Pivot>
  <Opacity value="100"/>
 </Transform>
 <Fill type="0"/>
 <Outline stroke-linejoin="round" stroke-linecap="round" type="1">
  <Color mode="0">
   <Red value="0"/>
   <Green value="0.40000000000000002"/>
   <Blue value="1"/>
   <Alpha value="1"/>
  </Color>
  <Width value="10"/>
 </Outline>
 <Paths>
  <Path mode="0" value="* * 0 0 * * 2,150.161 346.524 150.161 349.412 150.161 352.3 1" name="path" closed="false"/>
 </Paths>
</Object>
```

**Comment by kaixoo12** [comment]
We need to change how we save/load stuff. I don't like these `writeXEV`, `readXEV`, etc. functions at all.

**Comment by donn-xx** [comment]
(Newb) Please keep git in mind! Using binary files (like a zip) causes git to duplicate the entire file on every commit. If the save file is text, it will fit like a glove into a git workflow. Look at how Godot does it, with their tscene (text) files etc.

Or have a text option (perhaps ".tfdesign") for those who want to use repos.

**Comment by kaixoo12** [comment]
Will this project file spec be open-source? I could use it in my own editor
